ctapipe/io/hessiofilereader.py

```diff
@@ -0,0 +1,211 @@
+                data.mc.h_first_int = first_int
+                data.mc.shower_primary_id = \
+                    pyhessio.get_mc_shower_primary_id()
+
+                # mc run header data
+                data.mcheader.run_array_direction = \
+                    pyhessio.get_mc_run_array_direction()
+
+                # this should be done in a nicer way to not re-allocate the
+                # data each time (right now it's just deleted and garbage
+                # collected)
+
+                data.r0.tel.clear()
+                data.r1.tel.clear()
+                data.dl0.tel.clear()
+                data.dl1.tel.clear()
+                data.mc.tel.clear()  # clear the previous telescopes
+
+                for tel_id in tels_with_data:
+
+                    # event.mc.tel[tel_id] = MCCameraContainer()
+
+                    data.mc.tel[tel_id].dc_to_pe \
+                        = pyhessio.get_calibration(tel_id)
+                    data.mc.tel[tel_id].pedestal \
+                        = pyhessio.get_pedestal(tel_id)
+
+                    data.r0.tel[tel_id].adc_samples = \
+                        pyhessio.get_adc_sample(tel_id)
+                    if data.r0.tel[tel_id].adc_samples.size == 0:
+                        # To handle ASTRI and dst files
+                        data.r0.tel[tel_id].adc_samples = \
+                            pyhessio.get_adc_sum(tel_id)[..., None]
+                    data.r0.tel[tel_id].adc_sums = \
+                        pyhessio.get_adc_sum(tel_id)
+                    data.mc.tel[tel_id].reference_pulse_shape = \
+                        pyhessio.get_ref_shapes(tel_id)
+
+                    nsamples = pyhessio.get_event_num_samples(tel_id)
+                    if nsamples <= 0:
+                        nsamples = 1
+                    data.r0.tel[tel_id].num_samples = nsamples
+
+                    # load the data per telescope/pixel
+                    hessio_mc_npe = pyhessio.get_mc_number_photon_electron
+                    data.mc.tel[tel_id].photo_electron_image \
+                        = hessio_mc_npe(telescope_id=tel_id)
+                    data.mc.tel[tel_id].meta['refstep'] = \
+                        pyhessio.get_ref_step(tel_id)
+                    data.mc.tel[tel_id].time_slice = \
+                        pyhessio.get_time_slice(tel_id)
+                    data.mc.tel[tel_id].azimuth_raw = \
+                        pyhessio.get_azimuth_raw(tel_id)
+                    data.mc.tel[tel_id].altitude_raw = \
+                        pyhessio.get_altitude_raw(tel_id)
+                    data.mc.tel[tel_id].azimuth_cor = \
+                        pyhessio.get_azimuth_cor(tel_id)
+                    data.mc.tel[tel_id].altitude_cor = \
+                        pyhessio.get_altitude_cor(tel_id)
+                yield data
+                counter += 1
+
+                if self.max_events and counter >= self.max_events:
```

Comment:
I think the base class should entirely deal with this whole `max_events` matter, so that explicit child classes do not have to bother about this, and devs to not have to repeat this little if in each `self._generator()` method. 

And I think it is possible. I will make a PR tomorrow.

Reply:
No need, I've made this change